kafkazk/partitions.go

```diff
@@ -75,6 +75,30 @@ func (p PartitionList) SortBySize(m PartitionMetaMap) {
 	sort.Sort(partitionsBySize{pl: p, pm: m})
 }
 
+// replicasByLeaderFollowerRatio is used to shuffle replica
+// sets according to the broker leader to follower ratio.
+type replicasByLeaderFollowerRatio struct {
+	replicas []int
+	stats    BrokerUseStatsMap
+}
+
+func (r replicasByLeaderFollowerRatio) Len() int { return len(r.replicas) }
+
+func (r replicasByLeaderFollowerRatio) Swap(i, j int) {
+	r.replicas[i], r.replicas[j] = r.replicas[j], r.replicas[i]
+}
+
+func (r replicasByLeaderFollowerRatio) Less(i, j int) bool {
+	id1 := r.replicas[i]
+	id2 := r.replicas[j]
+
+	if r.stats[id1].Follower == 0 || r.stats[id2].Follower == 0 {
```

Comment:
I'm not sure this condition is right:
- if the first has no followers, then return false (because the ratio is infinite)
- if the second has no followers, then return true (because the ratio is infinite)
- if none of them have followers, then return what you return right now

```
if r.stats[id1].Follower == 0 && r.stats[id2].Follower == 0 {
    return r.stats[id1].Leader > r.stats[id2].Leader
}
if r.stats[id1].Follower == 0 {
    return false
}
if r.stats[id2].Follower == 0 {
    return true
}
```

not sure if it really matters in practice

Reply:
I actually tested a few variations like this and found it to not make a difference in the final results, but I did want to think about it a bit more.

Reply:
Thinking about this more, it could make a difference in partitions with replication factors >2 (as the replication factor increases, the set size increases, leaving more room for optimizations).

Reply:
Some follow up:

Realized I needed to switch the compare to a `<` in this branch anyway. Created another branch that includes more elaborate conditionals: https://github.com/DataDog/kafka-kit/compare/jamie/rebuild-leadership...jamie/rebuild-leadership-algo2.

Found that in comparing the current (with the corrected equality) and the algo2 branch that the current occasionally results in more optimal ratios:

`algo2` branch:
```
$ ./topicmappr rebuild --topics points-nov18 --brokers -1 --optimize-leaders | grep 1673
  points-nov18 p7: [1673 1796] -> [1673 1796] no-op
  points-nov18 p66: [1801 1673] -> [1673 1801] preferred leader
  points-nov18 p67: [1797 1673] -> [1673 1797] preferred leader
  Broker 1673 - leader: 3, follower: 0, total: 3
```

Current branch:
```
$ ./topicmappr2 rebuild --topics points-nov18 --brokers -1 --optimize-leaders | grep 1673             
  points-nov18 p7: [1673 1796] -> [1673 1796] no-op
  points-nov18 p66: [1801 1673] -> [1673 1801] preferred leader
  points-nov18 p67: [1797 1673] -> [1797 1673] no-op
  Broker 1673 - leader: 2, follower: 1, total: 3
```